crates/generated_parser/src/early_errors.rs

```diff
@@ -6,6 +6,61 @@ use ast::arena;
 use ast::source_atom_set::{CommonSourceAtomSetIndices, SourceAtomSet, SourceAtomSetIndex};
 use std::collections::HashMap;
 
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum ControlKind {
+    Continue,
+
+    Break,
+}
+
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub struct ControlInfo {
+    pub label: Option<SourceAtomSetIndex>,
+    // The offset of the nested control in the source.
+    pub offset: usize,
+    pub kind: ControlKind,
+}
+
+impl ControlInfo {
+    pub fn new_continue(offset: usize, label: Option<SourceAtomSetIndex>) -> Self {
+        Self {
+            label,
+            kind: ControlKind::Continue,
+            offset,
+        }
+    }
+
+    pub fn new_break(offset: usize, label: Option<SourceAtomSetIndex>) -> Self {
+        Self {
+            label,
+            kind: ControlKind::Break,
+            offset,
+        }
+    }
+}
+
+pub struct BreakOrContinueIndex {
+    pub index: usize,
+}
+
+impl BreakOrContinueIndex {
+    fn new(index: usize) -> Self {
+        Self { index }
+    }
+}
+
+impl From<BreakOrContinueIndex> for usize {
+    fn from(index: BreakOrContinueIndex) -> usize {
+        index.index
+    }
+}
+
+impl From<usize> for BreakOrContinueIndex {
+    fn from(index: usize) -> BreakOrContinueIndex {
+        BreakOrContinueIndex::new(index)
+    }
+}
```

Comment:
conversion from usize (untyped) to BreakOrContinueIndex (typed) should be explicit about the name,
otherwise the caller can convert unrelated thing unexpectedly
so IMO this `From` shouldn't exist.